tools/updater/main.c

```diff
@@ -91,99 +91,130 @@ static void print_supprted_commands()
     printf("    %s -Update c:\\data\\firmware.bin 0123456\n", EXECUTABLE_NAME);
 }
 
-static k4a_result_t ensure_firmware_open(updater_command_info_t *command_info);
-static void command_list_devices(void);
+static k4a_result_t ensure_firmware_open(updater_command_info_t *command_info, uint32_t device);
+static void command_list_devices(updater_command_info_t *command_info);
 static k4a_result_t command_query_device(updater_command_info_t *command_info);
 static k4a_result_t command_inspect_firmware(updater_command_info_t *command_info);
 static k4a_result_t command_update_device(updater_command_info_t *command_info);
 static k4a_result_t command_reset_device(updater_command_info_t *command_info);
+static void close_all_handles(updater_command_info_t *command_info);
 
-static int try_parse_device(int argc, char **argv, int current, updater_command_info_t *command_info)
+static k4a_result_t add_device(updater_command_info_t *command_info, char *serial_number)
 {
-    int next = current + 1;
-    if (next >= argc)
+    for (uint32_t index = 0; index < command_info->device_count; index++)
     {
-        // There is no more arguments
-        return current;
-    }
-
-    char firstCharacter = argv[next][0];
-    if (firstCharacter == '-' || firstCharacter == '/')
-    {
-        // The next argument is a new flag.
-        return current;
+        if (strcmp(serial_number, command_info->device_serial_number[index]) == 0)
+        {
+            // device present in list;
+            return K4A_RESULT_FAILED;
+        }
     }
 
-    char *device_serial_number = argv[next];
+    // Save the serial number.
+    command_info->device_serial_number[command_info->device_count] = serial_number;
+    command_info->device_count++;
+    return K4A_RESULT_SUCCEEDED;
+}
 
-    // NOTE: Leaving this as depthmcu since we only need access to the serial number and firmware will block until the
-    // device is ready for other commands that aren't needed.
-    depthmcu_t device = NULL;
+static int try_parse_device(int argc, char **argv, int current, updater_command_info_t *command_info)
+{
+    bool device_found = false;
+    depthmcu_t depth = NULL;
+    colormcu_t color = NULL;
     uint32_t device_count;
-    usb_cmd_get_device_count(&device_count);
-
-    command_info->device_index = UINT32_MAX;
+    char *device_serial_number = NULL;
+    int next = current + 1;
 
-    for (uint32_t deviceIndex = 0; deviceIndex < device_count; deviceIndex++)
+    if (next < argc)
     {
-        if (K4A_RESULT_SUCCEEDED != depthmcu_create(deviceIndex, &device))
+        char firstCharacter = argv[next][0];
+        if (firstCharacter != '-' && firstCharacter != '/')
         {
-            printf("ERROR: %d: Failed to open device\n", deviceIndex);
-            continue;
+            // If the user is passing in a serial number, then we use it and exit. We will later check for it
+            device_serial_number = argv[next];
         }
```

Comment:
I did this in an earlier iteration, but we need to validate the target serial number is present and fail out here rather than let ensure_firmware_open attempt to open it the first time as it has retry logic and will take ~30s to faile.